### ISSUE TITLE
CustomPropertyComparator should be used compare Collection items

### ISSUE THREAD

**Issue by bartoszwalacik** [javers/javers#747]
Currently, `CustomPropertyComparator` is used only for comparing object properties, for example, it can be used to compare ` Person` objects in `boss` property:

```java
class Company {
        private Person boss
        private List<Person> partners
}
```

But it is not used by JaVers to compare ` Person` objects when they are collection items in the `partners` List.

Possible solution is a new method in `CustomPropertyComparator` with default impl:

```java
   /**
     * This comparator is called by JaVers to calculate collection-to-collection diff.
     * <br/>
     * For example, when objects of type T are List items.
     * <br/><br/>
     * 
     * Both equals() and compare() should return consistent results. When compare() returns null,
     * equals() should return false.
     * 
     */
    default boolean equals(T a, T b) {
        return Objects.equals(a, b);
    }
```

**Comment by bartoszwalacik** [comment]
released in 4.0.0-RC3

**Comment by baumgartner** [comment]
```
    @Test
    public void equalEntityClientTest() {
        Person e1 = new Person("james", "ignore this");
        List<Person> l1 = new ArrayList<Person>();
        l1.add(e1);
        Company le1 = new Company("1", null, l1, null);

        Person e2 = new Person("james", "");
        List<Person> l2 = new ArrayList<Person>();
        l2.add(e2);
        Company le2 = new Company("1", null, l2, null);
        Diff diff = javers.compare(le1, le2);

        System.out.println(diff);
        assertEquals(0, diff.getChanges().size());
    } 
```


Works with ListCompareAlgorithm.SIMPLE and AS_SET, but with LEVENSTEIN_DISTANCE the following exception is thrown. 

```
java.lang.ClassCastException: java.util.Collections$UnmodifiableSet cannot be cast to java.util.List
	at org.javers.core.diff.appenders.levenshtein.LevenshteinListChangeAppender.calculateChanges(LevenshteinListChangeAppender.java:39)
	at org.javers.core.diff.appenders.levenshtein.LevenshteinListChangeAppender.calculateChanges(LevenshteinListChangeAppender.java:21)
	at org.javers.core.diff.appenders.CollectionAsListChangeAppender.calculateChanges(CollectionAsListChangeAppender.java:32)
	at org.javers.core.diff.appenders.CollectionAsListChangeAppender.calculateChanges(CollectionAsListChangeAppender.java:14)
	at org.javers.core.diff.appenders.CorePropertyChangeAppender.calculateChanges(CorePropertyChangeAppender.java:45)
	at org.javers.core.diff.DiffFactory.appendChanges(DiffFactory.java:151)
	at org.javers.core.diff.DiffFactory.appendPropertyChanges(DiffFactory.java:141)
	at org.javers.core.diff.DiffFactory.createAndAppendChanges(DiffFactory.java:123)
	at org.javers.core.diff.DiffFactory.create(DiffFactory.java:69)
	at org.javers.core.diff.DiffFactory.compare(DiffFactory.java:54)
	at org.javers.core.JaversCore.compare(JaversCore.java:173)
	at javers.test.JaversListTest.equalEntityClientTest(JaversListTest.java:37)
```

**Comment by baumgartner** [comment]
Works with ListCompareAlgorithm.SIMPLE
Assert fails for ListCompareAlgorithm.AS_SET and LEVENSHTEIN_DISTANCE. 
Test-Project available at https://github.com/baumgartner/javerstest/blob/master/src/test/java/javers/test/ListCompareTest.java

**Comment by bartoszwalacik** [comment]
omg, so we need a fix ...

**Comment by bartoszwalacik** [comment]
Fixed in 5.0.1 when CustomPropertyComparator is used with LEVENSHTEIN.
CustomPropertyComparator can't be used with AS_SET because for comparing Sets, we need both equals() and hashCode() and CustomPropertyComparator implements only custom equals()

**Comment by dmitry-weirdo** [comment]
@bartoszwalacik any perspective to also be able to use ` `CustomPropertyComparator` together with `CustomPropertyComparator`?

I described my case on https://github.com/javers/javers/issues/1038.
